src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs

```diff
@@ -5,130 +5,188 @@
+            }
+
+            _knownSources = new HashSet<string>();
+            foreach (var source in _sources.Keys)
+            {
+                _knownSources.Add(source);
             }
         }
 
-        private void ProtocolDiagnostics_Event(ProtocolDiagnosticEvent pdEvent)
+        private void ProtocolDiagnostics_ResourceEvent(ProtocolDiagnosticResourceEvent pdEvent)
         {
-            AddAggregateData(pdEvent, _data);
+            AddResourceData(pdEvent, _data, _resourceStringTable, _knownSources);
         }
 
-        internal static void AddAggregateData(ProtocolDiagnosticEvent pdEvent, ConcurrentDictionary<string, Data> allData)
+        internal static void AddResourceData(
+            ProtocolDiagnosticResourceEvent pdEvent,
+            ConcurrentDictionary<string, Data> allData,
+            ConcurrentDictionary<string, ConcurrentDictionary<string, string>> resourceStringTable,
+            HashSet<string> sources)
         {
+            if (!sources.Contains(pdEvent.Source))
+            {
+                return;
+            }
+
+            var resourceMethodNameTable = resourceStringTable.GetOrAdd(pdEvent.ResourceType, t => new ConcurrentDictionary<string, string>());
+            var resourceTypeAndMethod = resourceMethodNameTable.GetOrAdd(pdEvent.Method, m => pdEvent.ResourceType + "." + m);
+
             var data = allData.GetOrAdd(pdEvent.Source, _ => new Data());
+            lock (data._lock)
+            {
+                if (data.Resources.TryGetValue(resourceTypeAndMethod, out var t))
+                {
+                    data.Resources[resourceTypeAndMethod] = (t.count + 1, t.duration + pdEvent.Duration);
+                }
+                else
+                {
+                    data.Resources[resourceTypeAndMethod] = (1, pdEvent.Duration);
+                }
+            }
+        }
 
-            var resourceData = pdEvent.Url.OriginalString.EndsWith(".nupkg", StringComparison.OrdinalIgnoreCase)
-                ? data.Nupkg
-                : data.Metadata;
+        private void ProtocolDiagnostics_HttpEvent(ProtocolDiagnosticHttpEvent pdEvent)
+        {
+            AddHttpData(pdEvent, _data, _knownSources);
+        }
 
-            lock (resourceData.Lock)
+        internal static void AddHttpData(ProtocolDiagnosticHttpEvent pdEvent, ConcurrentDictionary<string, Data> allData, HashSet<string> sources)
+        {
+            if (!sources.Contains(pdEvent.Source))
             {
-                ApplyTiming(resourceData.EventTiming, pdEvent.EventDuration);
+                return;
+            }
 
-                if (pdEvent.HeaderDuration.HasValue)
-                {
-                    if (resourceData.HeaderTiming == null)
-                    {
-                        resourceData.HeaderTiming = new ResourceTimingData();
-                    }
+            var data = allData.GetOrAdd(pdEvent.Source, _ => new Data());
```

Comment:
Shouldn't it just be Get and NOT Add? If something gets added here, that means your sources (`_knownSources`) is now out of sync with allData (`_data`).

Reply:
`_data` wasn't pre-populated with data for all known sources, so `GetOrAdd` was needed. But then I realized I could just pre-populate it with all known sources, and remove the `_knownSources` hash set.